basics_of_sst-core/tutorial/Example04.md

```diff
@@ -0,0 +1,488 @@
+
+There are other ELI macros that may be necessary depending upon circumstances.  For example, if statistics are defined for the subcomponent they must be documented using the `SST_ELI_DOCUMENT_STATISTICS` macro.  However, for this simple case, the two reference above will be sufficient.
+
+Finally, a single state variable is defined for use by the subcomponent.
+
+```
+        private:
+            int count_ = 0;
+```
+
+## Update the Component
+
+Open the file `ExampleComponent.h` and add a reference to the subcomponent and ExampleSubComponent include files.
+
+```
+#include <sst/core/component.h>
+#include <sst/core/subcomponent.h>
+#include <sst/core/link.h>
+#include <sst/core/output.h>
+#include <sst/core/params.h>
+
+#include "ExampleSubComponent.h"
+```
+
+Then add a public variable to hold the subcomponents.  Similar to links, subcomponents are stored in subcomponent slots.  The difference is, each slot can hold multiple subcomponents so a vector of subcomponents is created rather than a single subcomponent.   
+
+```
+            // Links to connect to other component.
+            // The link connects to the components port.
+            //
+            SST::Link* link_;
+
+            // Slot to hold a subcomponent.
+            //
+            std::vector<ExampleSubComponent*> subcomponentSlots_;
+```
+
+Subcomponent slots are registered with SST via the `ELI SST_ELI_DOCUMENT_SUBCOMPONENT_SLOTS` macro, so insert the following code below the `SST_ELI_DOCUMENT_PORTS` macro.
+
+```
+            // Document the subcomponent slots.
+            //
+            SST_ELI_DOCUMENT_SUBCOMPONENT_SLOTS(
+                // Slot name, description, slot type
+                //
+                // Slot name is just a name.  It can be anything that makes sense but will
+                // be used later to refer to this slot.
+                //
+                // Description is just that.  Can be anything.
+                //
+                // Slot type is the subcomponent fully qualified class name.
+                //
+                {"slot_", "Slot to hold a subcomponent", "Example04::ExampleSubComponent"}
+            )
+```
+
+This registers a subcomponent named `"slot_"` with a description of `"Slot to hold a subcomponent"` and type `"Example04::ExampleSubComponent"` with SST so it can be used by the simulation.  
+
+Save this file and open the file `ExampleComponent.cc`.  Insert the following code to configure the subcomponent into the constructor.
+
+```
+    // Configure the subcomponents.  The slot name is taken from where the 
+    // subcomponent slot is documented in the component .h file.
+    //
+    SubComponentSlotInfo* info = getSubComponentSlotInfo("slot_");
+    if ( !info ) 
+    {
+        // No defined subcomponents.  Print an error message and exit.
+        //
+        Output::getDefaultObject().fatal(CALL_INFO, -1, "Must specify at least one SubComponent for slot.\n");
```

Comment:
Better practice to use the Component's fatal() function if you aren't going to use the component's own output object. This also will append info about the component to it. So: `fatal(CALL_INFO, exit_code, "message");` Otherwise you could call "getSimulationOutput()" to get an output object and call fatal() on that.

Reply:
I want to emphasize this.  BaseComponent::fatal() is relatively new, but adds a lot more information to the output than calling Output::fatal().  We should be encouraging developers to use that function for fatal errors instead of an Output object of any sort inside of a Component or SubComponent.